### PR TITLE
feat(pedm): add base PEDM implementation

### PR DESCRIPTION
**Pull request by kbouchard-dev**
CI changes:
- ~~Environment variable `DAGENT_PEDM_PFX` should be defined as an absolute path to a code signing certificate to sign the `Appx` package for the right click context menu (Windows 11 requires this to avoid going into the 'Show more options' option)~~
- ~~Environment variable `DAGENT_PEDM_PFX_PASSWORD` should be defined as the password of the certificate. **This is bad**, I would like to discuss other ways of doing it but I lack experience in CI/CD secret handling.~~
- ~~CI/CD .yml files will need to be updated.~~ Done

OpenAPI:
- The OpenAPI spec for the PEDM is generated with [`aide`](https://github.com/tamasfe/aide/). I noticed the repo uses another library and so if we should use only one I'd be glad to change it.
- Previously the script to generate the OpenAPI specs used the PowerShell `>` operator, which will write files in UTF16, which is a bit cumbersome to use. I replaced it with `| Out-File -Encoding UTF8`

Enabling the feature:
- ~~For the moment, you will need to toggle Devolutions PEDM in the installer as well as modify the file `C:\ProgramData\Devolutions\Agent\agent.json` to include `"Pedm":{"Enabled":true}` in the root object.~~
  - ~~To do this, you will need Administrator privileges~~
- Toggle 'PEDM' in the installer.

Known issues:
- ~~The virtual account feature is a bit unstable. Over RDP, applications sometimes entirely fail to show UI, and not with RDP, the applications load an old themed UI. This is probably due to the virtual account not having enough permissions in the actual user's desktop and session. Furthermore, there are probably issues linked to permissions and logging in to third parties because of this different user.~~ Fixed
  - See: https://www.reddit.com/r/Intune/comments/16z2uue/epm_network_share/
  - Worth investigating if this feature is needed
- Currently a process requesting changes to the PEDM configuration with an elevated token will always succeed. This also includes tokens created by the PEDM, and so an elevation can lead to full elevation everywhere. A possible way to fix this issue is "tagging" each token created by the PEDM and rejecting them when trying to modify the settings. This would be done though a token's attributes. However, we must verify that the attributes cannot be changed.
- The PEDM does not currently support elevating to medium integrity or elevating to high integrity. Right now, all elevations are in high integrity.
- The PEDM does not currently support changing behaviors for child processes. This is possible but not implemented yet.
- The `Appinfo` hook is in the code, but not enabled.

### CODE REVIEW

**Comment by pacmancoder** [comment]
@thenextman could you please review the wix installer changes in this PR please?

**Comment by thenextman** [comment]
> @thenextman could you please review the wix installer changes in this PR please?

I went through this installer and everything looks good, nice job 🤙 

My only concern is that we seem to be shipping a .NET 8 application but there's no provision for checking that dependency at install time. It does mean that if the runtime is not installed, the user will be prompted to install that the first time it's run which perhaps isn't a great experience (as we've seen with RDM).

I don't see it needing to be fixed as part of this PR, but it should be addressed before this goes into the user hands (whether that be "properly" by applying the dependency via the installer, like RDM by doing it inside the installer but not using Windows Installer features, simply blocking the install until the dependency is there, or something else).

**Comment by kbouchard-dev** [comment]
> > @thenextman could you please review the wix installer changes in this PR please?
> 
> I went through this installer and everything looks good, nice job 🤙
> 
> My only concern is that we seem to be shipping a .NET 8 application but there's no provision for checking that dependency at install time. It does mean that if the runtime is not installed, the user will be prompted to install that the first time it's run which perhaps isn't a great experience (as we've seen with RDM).
> 
> I don't see it needing to be fixed as part of this PR, but it should be addressed before this goes into the user hands (whether that be "properly" by applying the dependency via the installer, like RDM by doing it inside the installer but not using Windows Installer features, simply blocking the install until the dependency is there, or something else).
> 
> Note that this becomes especially important if this will be deployed by MDM (e.g. InTune). The .NET dependency cannot be easily installed by the MDM and should be handled on our side.

Thanks and good catch. As there's nothing too fancy going on I'll see if it's possible to downgrade to .NET Framework as to make it compatible by default.

**Comment by thenextman** [comment]
> > > @thenextman could you please review the wix installer changes in this PR please?
> > 
> > 
> > I went through this installer and everything looks good, nice job 🤙
> > My only concern is that we seem to be shipping a .NET 8 application but there's no provision for checking that dependency at install time. It does mean that if the runtime is not installed, the user will be prompted to install that the first time it's run which perhaps isn't a great experience (as we've seen with RDM).
> > I don't see it needing to be fixed as part of this PR, but it should be addressed before this goes into the user hands (whether that be "properly" by applying the dependency via the installer, like RDM by doing it inside the installer but not using Windows Installer features, simply blocking the install until the dependency is there, or something else).
> > Note that this becomes especially important if this will be deployed by MDM (e.g. InTune). The .NET dependency cannot be easily installed by the MDM and should be handled on our side.
> 
> Thanks and good catch. As there's nothing too fancy going on I'll see if it's possible to downgrade to .NET Framework as to make it compatible by default. For now it would work with RDM, but the error dialog can be shown when a user is not allowed to elevate and it's not guaranteed RDM is installed for that popup.

@kbouchard-dev That would be awesome (downgrade to .NET Framework) if possible. It also avoids issues with multi arch and simplifies the build.

**Comment by kbouchard-dev** [comment]
For the outdated `tower` and `hyper` dependencies in `devolutions-pedm-shared`, this should be fixed with the 7.8.0 release of OpenAPI Generator: https://github.com/OpenAPITools/openapi-generator/pull/19115

**Comment by pacmancoder** [comment]
Just a heads-up, I am currently in the process of reviewing this PR, optimistically, I'll be able to finish it tomorrow

File: crates/devolutions-pedm/openapi/generate_all.ps1
Comment (pacmancoder) [inline crates/devolutions-pedm/openapi/generate_all.ps1:8]: I believe its better to write path relative to script root ($PSScriptRoot variable) isntead of relative path (`../../..`)
cc @awakecoding

File: crates/devolutions-pedm/src/config.rs
Comment (pacmancoder) [inline crates/devolutions-pedm/src/config.rs:9]: Is this TODO should be removed or renaming should occur?

File: crates/devolutions-pedm/src/elevator/local_admin_elevator.rs
Comment (pacmancoder) [inline crates/devolutions-pedm/src/elevator/local_admin_elevator.rs:48]: leftover comment?

File: crates/devolutions-pedm/src/error.rs
Comment (pacmancoder) [inline crates/devolutions-pedm/src/error.rs:18]: maybe lets use `thiserror` here?

File: crates/devolutions-pedm/src/utils.rs
Comment (pacmancoder) [inline crates/devolutions-pedm/src/utils.rs:70]: Why do we actually need multi hash? is it required for WinAPI calls or for the other reason?

File: crates/win-api-wrappers/src/error.rs
Comment (pacmancoder) [inline crates/win-api-wrappers/src/error.rs:18]: Might be good to use `thiserror` here

File: crates/win-api-wrappers/src/lib.rs
Comment (pacmancoder) [inline crates/win-api-wrappers/src/lib.rs:6]: Lets add here comment about reason for silencing lints with `allow`

File: crates/win-api-wrappers/src/rpc.rs
Comment (pacmancoder) [inline crates/win-api-wrappers/src/rpc.rs:91]: It would be great to wrap this  in some kind of wrapper, which calls RpcRevertToSelfEx on destructor

File: crates/win-api-wrappers/src/win.rs
Comment (pacmancoder) [inline crates/win-api-wrappers/src/win.rs:1]: Lets split this file to multiple small modules, single with >3K lines of code is very hard to follow 😢

File: package/AgentWindowsManaged/Actions/AgentActions.cs
Comment (pacmancoder) [inline package/AgentWindowsManaged/Actions/AgentActions.cs:214]: Indeed sounds like a very strange behavior

**Review by pacmancoder — COMMENTED** [review]
I gave the PR a first review round, it is a very impressive chunk of work!

File: ci/tlk.ps1
Comment (thenextman) [reply ci/tlk.ps1:464]: Yes, signing needs to be done in the CI workflow. You can invoke devops to get the proper certificates/secrets made available. 

It's ok to allow signing in tlk for dev/test purposes with a local certificate, but not for releases.

Signing is normally only done in the package workflow since privileged access to the signing information requires the workflow run to be authorized. If it's important that CI artifacts are also signed, we'd need to consider options here (perhaps sign in CI with a dev / low privilege certificate, and then resign in packaging)

File: ci/tlk.ps1
Comment (thenextman) [reply ci/tlk.ps1:389]: Minor correction; it's the package workflow that will call tlk. The release workflow just pushes artifacts / creates the release. Bur yes, someone on architecture-maintainers will need to authorize a workflow run.

File: crates/devolutions-pedm-contextmenu/devolutions-pedm-contextmenu-shell/src/lib.rs
Comment (thenextman) [reply crates/devolutions-pedm-contextmenu/devolutions-pedm-contextmenu-shell/src/lib.rs:154]: Viewing this completely out of context, but are we assuming Windows is installed on the C: drive?

File: crates/devolutions-pedm-shared/devolutions-pedm-client-http/.gitignore
Comment (kbouchard-dev) [reply crates/devolutions-pedm-shared/devolutions-pedm-client-http/.gitignore:1]: This is autogenerated by OpenAPI Generator. We could add a .gitignore for files it generates. I'll see if there are configurations to change its behavior.

File: crates/devolutions-pedm/src/utils.rs
Comment (kbouchard-dev) [reply crates/devolutions-pedm/src/utils.rs:70]: Not really. It's more of a convenience feature. If the user wants to add a hash check they can select either sha256 or sha1. There's sha1 since it's popular for certificate fingerprints. This is just a helper that computes all the supported hashes as one.

File: crates/devolutions-pedm-contextmenu/devolutions-pedm-contextmenu-shell/src/lib.rs
Comment (pacmancoder) [reply crates/devolutions-pedm-contextmenu/devolutions-pedm-contextmenu-shell/src/lib.rs:154]: @thenextman huh, is it possible to install to non-C drive? 😮

Comment (thenextman) [reply crates/devolutions-pedm-contextmenu/devolutions-pedm-contextmenu-shell/src/lib.rs:154]: @pacmancoder It's possible that there isn't even a C drive

One way to know would be to call `SHGetKnownFolderPath` with `FOLDERID_Windows` (in C, at least :))

File: ci/tlk.ps1
Comment (CBenoit) [reply ci/tlk.ps1:389]: I started a dry run on this branch to catch potential errors.

File: crates/devolutions-pedm-hook/src/hook.rs
Comment (CBenoit) [reply crates/devolutions-pedm-hook/src/hook.rs:8]: @kbouchard-dev You can automatically format using this style by running rustfmt with the nightly toolchain.

```bash
cargo +nightly fmt
```

File: crates/devolutions-pedm-shared/Cargo.toml
Comment (pacmancoder) [reply crates/devolutions-pedm-shared/Cargo.toml:28]: Ah, I am sorry, my comment was a little bit confusing, I wanted to say that instead of empty `build` feature set, there should be an empty `default` feature set, that is more in-line with rest of our code

File: ci/tlk.ps1
Comment (CBenoit) [reply ci/tlk.ps1:389]: @kbouchard-dev 
The CI appears broken from start. Just in case, can you rebase on master and push again?

File: crates/devolutions-pedm-shared/Cargo.toml
Comment (CBenoit) [reply crates/devolutions-pedm-shared/Cargo.toml:28]: The `build` feature is actually for feature-gating a module called `build`, so my understanding is that we can’t replace it with `default` :thinking:

File: crates/devolutions-pedm-shared/Cargo.toml
Comment (CBenoit) [inline crates/devolutions-pedm-shared/Cargo.toml:3]: suggestion: Internal crates that we don’t publish (with `publish = false`), are assigned the version `0.0.0` because it doesn’t really matter and we don’t want to keep updating that number at each release.

File: crates/devolutions-pedm-shared/devolutions-pedm-client-http/.gitignore
Comment (CBenoit) [reply crates/devolutions-pedm-shared/devolutions-pedm-client-http/.gitignore:1]: You can add a .openapi-generator-ignore file for ignoring files generated by the generator that we don’t need: https://github.com/Devolutions/devolutions-gateway/blob/master/devolutions-gateway/openapi/dotnet-client/.openapi-generator-ignore

**Comment by CBenoit** [comment]
I’m not sure why, but the CI appears to not be running on your PR. I tried to remove the draft status in case that’s the cause (it’s probably not a draft anymore anyway).

File: crates/devolutions-pedm/openapi/generate_all.ps1
Comment (CBenoit) [reply crates/devolutions-pedm/openapi/generate_all.ps1:8]: It’s fine, because there is a `Push-Location -Path $PSScriptRoot` at the top of the file, ensuring we are always relative to the file.

File: crates/devolutions-pedm-shared/devolutions-pedm-client-http/src/models/error.rs
Comment (CBenoit) [inline crates/devolutions-pedm-shared/devolutions-pedm-client-http/src/models/error.rs:37]: suggestion: If you only want to output the name of the variant instead of a pretty message, you can simply to this:
```suggestion
        write!(f, "{self:?}")
```

File: crates/devolutions-pedm/src/error.rs
Comment (CBenoit) [reply crates/devolutions-pedm/src/error.rs:16]: In this case, I believe the benefit to be too minor to be worth it. This error is useful for deriving the `Display` trait with nice user-facing errors, but we don’t have such error messages here. See [this comment](https://github.com/Devolutions/devolutions-gateway/pull/942/files/0deddb6d97397eba1462f66fc044bff5f7d2e605#r1704158166).

File: ci/tlk.ps1
Comment (kbouchard-dev) [reply ci/tlk.ps1:389]: Just rebased

**Comment by kbouchard-dev** [comment]
CI bug looks like the Agent not being gated behind Windows properly. Will look into it.

**Review by CBenoit — COMMENTED** [review]
Here is the course of action I suggest:

- Document the unsafe blocks, explaining preconditions and why the usage is correct.
- Document and checkout important knowledge directly in committed README / modules files. This includes things like known issues, future plans, and / or a general overview of the module. I leave you deciding what is relevant, but for instance: what are the two elevation methods, and why we don’t yet use one of them (virtual account elevator).
- Enable the workspace lints: https://github.com/Devolutions/devolutions-gateway/blob/873217c8042167bf378613f812b3db2a399d256a/jetsocat/Cargo.toml#L9-L10
  - If that’s too much work, just enable the "safer unsafe" lints listed here: https://github.com/Devolutions/devolutions-gateway/blob/873217c8042167bf378613f812b3db2a399d256a/Cargo.toml#L31-L59
  - We can help you enabling remaining lints and fixing them in follow up PRs.
- Dry run of `Create New Release` on the branch to see if the workflow is still fine.

I think the PR can be merged once all of this is addressed, and we can follow up on it in separate PRs to help Kyle.

I’ll be on vacation starting tomorrow, can you take care of landing the PR? @pacmancoder

**Comment by pacmancoder** [comment]
@kbouchard-dev ping me when PR will be available for re-review 👀

File: .github/workflows/package.yml
Comment (CBenoit) [inline .github/workflows/package.yml:158]: nitpick: Feel free to remove unused code.

**Comment by CBenoit** [comment]
Package workflow is failed with this error:

image

File: .github/workflows/package.yml
Comment (kbouchard-dev) [reply .github/workflows/package.yml:158]: This was because the MacOS runner in test mode didn't have proper keys for signing. Will update it now now that the agent packages correctly.

**Review by rbstp — APPROVED** [review]
CI portion looks good

**Pull request merged**
